### PR TITLE
MCR-3101 use MCRScopedSession to perform in thread acl checks

### PR DESCRIPTION
**Pull request by yagee-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3101).

### CODE REVIEW

File: mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java [hunk 1]
```diff
@@ -426,33 +415,12 @@ public static boolean hasRule(String id, String permission) {
...
-                LOGGER.error("Exception while running ACL check for user: {}", user.getUserID(), e);
-                return false;
-            }
-        };
+    public static boolean checkPermission(MCRUserInformation user, Supplier<Boolean> checkSuplier) {
+        if (!MCRSessionMgr.hasCurrentSession()
+            || !(MCRSessionMgr.getCurrentSession() instanceof MCRScopedSession session)) {
+            throw new IllegalStateException("require an instance of MCRScopedSession");
+        }
+        return session.doAs(new MCRScopedSession.ScopedValues(user, "0.0.0.0"), checkSuplier);
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java]: Don't use magic string. Add getDefaultIP() to ScopedValues().

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.mycore.common.config.MCRConfiguration2;
+
+public final class MCRScopedSession extends MCRSession {
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java:38]: Missing javadoc.

File: mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java [hunk 3]
```diff
@@ -426,33 +415,12 @@ public static boolean hasRule(String id, String permission) {
...
-        MCRFixedUserCallable<Boolean> mcrFixedUserCallable = new MCRFixedUserCallable<>(check::get, user);
-        return () -> {
-            try {
-                return mcrFixedUserCallable.call();
-            } catch (Exception e) {
-                LOGGER.error("Exception while running ACL check for user: {}", user.getUserID(), e);
-                return false;
-            }
-        };
+    public static boolean checkPermission(MCRUserInformation user, Supplier<Boolean> checkSuplier) {
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java]: typo in checkSuplier

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    private ThreadLocal<ScopedValues> scopedValues = new ThreadLocal<>();
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java]: Can be final.

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        return values.userInformation();
+    }
+
+    @Override
+    public Iterator<Object> getObjectsKeyList() {
+        ScopedValues values = scopedValues.get();
+        if (values == null) {
+            return super.getObjectsKeyList();
+        }
+        return values.map.keySet().stream().collect(Collectors.toSet()).iterator();
```
Comment (Mewel) [inline mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java]: Can be return values.map.keySet().iterator();

**Review by Mewel — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java (hunk 1 above)
Comment (yagee-de) [reply mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java]: fixed in https://github.com/MyCoRe-Org/mycore/pull/2162/commits/231740ee2ac7ce66f4c60aa6131445f534ff1486

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java (hunk 2 above)
Comment (yagee-de) [reply mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java:38]: fixed in https://github.com/MyCoRe-Org/mycore/pull/2162/commits/231740ee2ac7ce66f4c60aa6131445f534ff1486

File: mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java (hunk 3 above)
Comment (yagee-de) [reply mycore-base/src/main/java/org/mycore/access/MCRAccessManager.java]: fixed in https://github.com/MyCoRe-Org/mycore/pull/2162/commits/231740ee2ac7ce66f4c60aa6131445f534ff1486

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java (hunk 4 above)
Comment (yagee-de) [reply mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java]: fixed in https://github.com/MyCoRe-Org/mycore/pull/2162/commits/231740ee2ac7ce66f4c60aa6131445f534ff1486

File: mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java (hunk 5 above)
Comment (yagee-de) [reply mycore-base/src/main/java/org/mycore/common/MCRScopedSession.java]: fixed in https://github.com/MyCoRe-Org/mycore/pull/2162/commits/231740ee2ac7ce66f4c60aa6131445f534ff1486

**Review by Mewel — APPROVED** [review]

**Comment by rsteph-de** [comment]
Shouldn't this be fixed in the 2023.06 LTS branch. If I remember correctly this fix would also improve the stability of the OAI interface.

**Comment by yagee-de** [comment]
> Shouldn't this be fixed in the 2023.06 LTS branch. If I remember correctly this fix would also improve the stability of the OAI interface.

There are workarounds available if the number of CPU threads is larger than the amount of connections in the connection pool. (use OAI-Prefix and `RunListRecordsParallel = false`)

**Pull request merged**